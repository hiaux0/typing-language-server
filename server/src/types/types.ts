import { Position } from "vscode-languageserver-textdocument";

/**
 * file://path/to/file.ext
 */
export type Uri = string;

/**
 * Type from the client: vscode.TextEditor.Selection
 */
export interface ClientEditorSelection {
  /**
   * The position at which the selection starts.
   * This position might be before or after [active](#Selection.active).
   */
  anchor: Position;
  /**
   * The position of the cursor.
   * This position might be before or after [anchor](#Selection.anchor).
   */
  active: Position;
  start: Position;
  end: Position;

  /**
   * A selection is reversed if [active](#Selection.active).isBefore([anchor](#Selection.anchor)).
   */
  // isReversed: boolean;
}

export interface GetEditorSelectionResponse {
  documentText: string;
  documentUri: string;
  documentPath: string;
  selections: ClientEditorSelection[];
}

export interface TypoAnalytics {
  text: string;
  mispelled: number;
}

export interface TypingAnalytics {
  occurrence: number;
  typos: TypoAnalytics[];
}

// export type AnalyticsMap = Map<string, Analytics>;
export type TypingAnalyticsMap = Record<string, TypingAnalytics>;

export type TypingLessons =
  | "words"
  | "alphabet" // abcdefghijklmnopqrstuvwxyz
  | "alphabet-chunks" // chunks of alphabet
  | "ac" // alphabet-chunks
  | "bigrams" // todo
  | "vim"; // todo
export interface WordsFilterConfigurationInput {
  lesson?: TypingLessons;
  amount: number; // amount of words
  length: number; // length of the word
  repeat?: number;
  autoEnter?: boolean; // Auto new line when finished. Note: Either this or clearLineOnFinish
  autoNewWords?: boolean;
  clearOnError?: boolean;
  clearLineOnFinish?: boolean; // Note: Either this or autoEnter
  // letters: string | string[], // "rs" or ["st", "rs"]
  ignore?: string | string[]; // "rs" or ["st", "rs"]
  anyOrder?: string | string[]; // "rs, at" or ["rs", "at"]
  oneOf?: string | string[]; // "rs" or ["r", "s"]
  sequence?: string | string[]; // "rs, at" or ["rs", "at"]
}

export interface WordsFilterConfigurationOutput
  extends Required<WordsFilterConfigurationInput> {
  ignore: string[]; // "rs" or ["st", "rs"]
  oneOf: string[]; // ["r", "s"]
  sequence: string[]; // ["r", "s"]
}
