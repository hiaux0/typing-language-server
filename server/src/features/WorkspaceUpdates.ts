import { Connection, RequestType } from 'vscode-languageserver';
import { GetEditorSelectionResponse } from '../types/types';
import {
	DocumentUri,
	OptionalVersionedTextDocumentIdentifier,
	Position,
	Range,
} from 'vscode-languageserver';
import {
	WorkspaceChange,
	ApplyWorkspaceEditResponse,
} from 'vscode-languageserver-protocol';
import { connection } from '../server';

// export async function getEditorSelection(connection: Connection) {
// 	const req = new RequestType('get-editer-selections');
// 	const getEditorResponse = (await connection.sendRequest(
// 		req
// 	)) as GetEditorSelectionResponse;
// 	return getEditorResponse;
// }

/**
 * from https://stackoverflow.com/questions/57787621/how-to-create-and-edit-a-new-file-in-the-workspace-via-language-server-extension
 */
export class WorkspaceUpdates {
	private _wschanges: WorkspaceChange;

	constructor() {
		this._wschanges = new WorkspaceChange();
	}

	public getEdits() {
		return this._wschanges.edit
	}

	hasChanges(): boolean {
		return (
			this._wschanges.edit.changes != undefined ||
			this._wschanges.edit.documentChanges != undefined
		);
	}

	createFile(uri: DocumentUri, contents: string, overwrite = false) {
		this._wschanges.createFile(uri, { overwrite: overwrite });
		const edit = this._wschanges.edit;
		const change = this._wschanges.getTextEditChange(
			OptionalVersionedTextDocumentIdentifier.create(uri, null)
		);
		this.insertText(uri, contents, 0, 0);
	}

	renameFile(uri: DocumentUri, newUri: DocumentUri, overwrite: boolean) {
		this._wschanges.renameFile(uri, newUri, { overwrite: overwrite });
	}

	deleteFileFolder(
		uri: DocumentUri,
		recursive: boolean,
		ignoreIfNotExists: boolean
	) {
		this._wschanges.deleteFile(uri, {
			recursive: recursive,
			ignoreIfNotExists: ignoreIfNotExists,
		});
	}

	insertText(uri: DocumentUri, contents: string, line: number, column: number) {
		const change = this._wschanges.getTextEditChange(
			OptionalVersionedTextDocumentIdentifier.create(uri, null)
		);
		change.insert(Position.create(line, column), contents);
	}

	replaceText(
		uri: DocumentUri,
		replaceWith: string,
		startLine: number,
		startColumn: number,
		endLine: number,
		endColumn: number
	) {
		const change = this._wschanges.getTextEditChange(
			OptionalVersionedTextDocumentIdentifier.create(uri, null)
		);
		change.replace(
			Range.create(
				Position.create(startLine, startColumn),
				Position.create(endLine, endColumn)
			),
			replaceWith
		);
	}

	replaceAllText(uri: DocumentUri, contents: string) {
		this.replaceText(uri, contents, 0, 0, Number.MAX_VALUE, Number.MAX_VALUE);
	}

	deleteText(
		uri: DocumentUri,
		contents: string,
		startLine: number,
		startColumn: number,
		endLine: number,
		endColumn: number
	) {
		const change = this._wschanges.getTextEditChange(
			OptionalVersionedTextDocumentIdentifier.create(uri, null)
		);
		change.delete(
			Range.create(
				Position.create(startLine, startColumn),
				Position.create(endLine, endColumn)
			)
		);
	}

	async applyChanges(): Promise<ApplyWorkspaceEditResponse> {
		return connection.workspace.applyEdit(this._wschanges.edit);
	}
}
