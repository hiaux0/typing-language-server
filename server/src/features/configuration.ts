import { getFencedCodeBlockContentNodeByName } from "../tree-sitter/ts-markdown"
import { WordsFilterConfigurationInput, WordsFilterConfigurationOutput } from "../types/types";

export const defaultFilterConfigurationOutput: WordsFilterConfigurationOutput = {
	lesson: "words",
	amount: 10,
	length: 999,
	repeat: 1,
	ignore: [],
	anyOrder: [],
	oneOf: [],
	sequence: [],
}

/**
 * @example
 * ```ts
 * const sourceCode = `
 * ```letters
 * rs
 * ``` 
 * `;
 * const result = getFilterLettersForWords(sourceCode);
 * result // => ["r", "s"]
 * ```
 */
export function getFilterConfigurationsForWords(sourceCode: string): WordsFilterConfigurationOutput {
	const node = getFencedCodeBlockContentNodeByName(sourceCode, "settings")
	const text = node?.node.text;
	if (!text) return defaultFilterConfigurationOutput;
	try {
		const asJson = JSON.parse(text) as WordsFilterConfigurationInput;
		const output = {
			// to keep the ordering of the properties
			...asJson,
			...defaultFilterConfigurationOutput
		}
		output.amount = asJson.amount;
		output.length = asJson.length;
		if (asJson.lesson) output.lesson = asJson.lesson;
		if (asJson.repeat) output.repeat = asJson.repeat;

		stringToArray("ignore")
		stringToArray("anyOrder")
		stringToArray("oneOf")
		stringToArray("sequence")

		return output;

		function stringToArray(key: keyof WordsFilterConfigurationOutput): void {
			if (typeof asJson[key] === 'string') {
				const value = asJson[key] as string
				const lettersArr = value
					.split(",")
					.map(char => char.trim());
				// @ts-ignore - output has props with number as keys, but we assign them at the start of the try block
				output[key] = lettersArr;
			}
		}
	} catch (error) {
		console.log("Not valid JSON");
		console.log("[configuration.ts,23] error: ", error);
		return defaultFilterConfigurationOutput;
	}
}
