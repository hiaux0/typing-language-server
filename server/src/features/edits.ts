import { WorkspaceUpdates } from "./WorkspaceUpdates";

export function getEdits() {
	const workspaceUpdates: WorkspaceUpdates = new WorkspaceUpdates();
	workspaceUpdates.applyChanges();
	// workspaceUpdates.replaceText(
	// 	documentUri,
	// 	withImports,
	// 	selection.start.line,
	// 	selection.start.character,
	// 	selection.end.line,
	// 	selection.end.character
	// );
	const edits = workspaceUpdates.getEdits();
	return edits;
}
