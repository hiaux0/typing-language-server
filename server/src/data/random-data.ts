import { defaultFilterConfigurationOutput } from '../features/configuration';
import { getRandomElement } from '../modules/array';
import { WordsFilterConfigurationOutput } from '../types/types';
import * as WordsData from './words.json'
// const WordsData = ["abc", "hello", "scream", "next", "right", "rest", "raise", "okay", "skim", "scry"]

type WordPool = Record<string, string[]>;

const wordsByLetter = WordsData.reduce((acc, word) => {
    const firstLetter = word[0];
    if (!acc[firstLetter]) {
        acc[firstLetter] = [];
    }
    acc[firstLetter].push(word);
    return acc;
}, {} as WordPool)



/**
 * A. amount
 * B. length
 * C. oneOf
 * D. ignore
 * E. sequence
 * F. repeat
 * G. anyOrder
 * 1. Distribution checker
 */
export function getRandomWords(amount: number = 10, filters?: WordsFilterConfigurationOutput): string[] {
    const finalFilters = {
        ...defaultFilterConfigurationOutput,
        ...filters,
    }
    /* F. */
    const finalAmount = finalFilters?.repeat ? Math.round(amount / finalFilters.repeat) : amount;
    // console.log("[random-data.ts,28] filters: ", filters);
    if (!finalFilters) {
        const len = WordsData.length;
        const rndArr = Array.from({ length: finalAmount }, () => Math.floor(Math.random() * len));
        const chosenWords = rndArr.map(i => WordsData[i]);
        return chosenWords;
    }

    let infiniteLoopCounter = 0;
    /* 1. */
    // To make sure each oneOf letter is present
    let currentOneOfIndex = 0;
    const wordCollector: Set<string> = new Set();
    /* C. */
    const oneOfFilter = finalFilters.oneOf ?? [];
    // Filter by selecting a random index, then from there find the next word, that matches the filter
    /* D. */
    const lettersIgnore = finalFilters.ignore ?? [];
    /* E. */
    const sequenceFilter = finalFilters.sequence ?? [];
    const wordLength = finalFilters.length
    const orderOfFilterProps = Object.keys(finalFilters);

    const orderingFilterFunctionMap: Record<string, Function> = {
        ignore: filterByIgnore,
        oneOf: filterByOneOf,
        length: filterByLength,
        sequence: filterBySequence
    };

    /* A. */
    // while ((wordCollector.size < finalAmount) && infiniteLoopCounter < 2) {
    while ((wordCollector.size < finalAmount) && infiniteLoopCounter < WordsData.length) {
        // console.log("------------------------------------------------------------");
        infiniteLoopCounter++;
        let wordPool = WordsData
        wordPool = wordPool.filter(word => !wordCollector.has(word));

        orderOfFilterProps.forEach(filterProp => {
            if (typeof orderingFilterFunctionMap[filterProp] !== 'function') return;
            wordPool = orderingFilterFunctionMap[filterProp](wordPool);
            // console.log("[random-data.ts,69] filterProp: ", filterProp);
            const sub = wordPool.slice(0, 20);
            // console.log("[random-data.ts,70] sub: ", sub);
        });

        const targetWord = getRandomElement(wordPool);
        if (!targetWord) {
            const popped = orderOfFilterProps.pop();
            continue;
        }
        if (wordCollector.has(targetWord)) continue;
        // console.log("[random-data.ts,75] targetWord: ", targetWord);
        wordCollector.add(targetWord);
    }
    const result = Array.from(wordCollector);
    if (finalFilters.repeat) return doubleResult(result);

    return result;

    function filterByOneOf(wordPool: string[]): string[] {
        /* C. */
        if (oneOfFilter.length === 0) return wordPool;
        const letter = oneOfFilter[currentOneOfIndex++ % oneOfFilter.length];
        wordPool = wordPool.filter(word => word.includes(letter));
        return wordPool;
    }
    function filterByLength(wordPool: string[]): string[] {
        /* B. */
        wordPool = wordPool.filter(word => word.length <= wordLength);
        return wordPool;
    }
    function filterByIgnore(wordPool: string[]): string[] {
        /* D. */
        if (lettersIgnore.length === 0) return wordPool
        wordPool = wordPool.filter(word => lettersIgnore.every(filter => !word.includes(filter)));
        return wordPool;
    }
    function filterBySequence(wordPool: string[]) {
        /* E. */
        if (sequenceFilter.length === 0) return wordPool;
        const sequence = sequenceFilter[currentOneOfIndex++ % sequenceFilter.length];
        wordPool = wordPool.filter(word => word.includes(sequence));
        return wordPool;
    }
    function doubleResult(result: string[]): string[] {
        const doubled: string[] = []
        result.forEach(word => {
            for (let i = 0; i < finalFilters.repeat; i++) {
                doubled.push(word)
            }
        })
        return doubled;
    }
}

