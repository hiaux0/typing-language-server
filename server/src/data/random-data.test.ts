import { defaultFilterConfigurationOutput } from "../features/configuration";
import { WordsFilterConfigurationOutput } from "../types/types";
import { getRandomWords } from "./random-data";

const ALPHABET = "abcdefghijklmnopqrstuvwxyz";
const baseConfig: WordsFilterConfigurationOutput =
  defaultFilterConfigurationOutput;
function getWordsFilterConfig(
  config: Partial<WordsFilterConfigurationOutput>,
): WordsFilterConfigurationOutput {
  return {
    ...baseConfig,
    ...config,
  };
}

describe("getRandomWords", () => {
  test("repeat", () => {
    const config = getWordsFilterConfig({
      amount: 6,
      sequence: ["sh", "ch"],
      length: 6,
      repeat: 3,
    });
    const result = getRandomWords(config.amount, config);
  });
  describe("ignore", () => {
    test("ignore - 1", () => {
      const config = getWordsFilterConfig({
        amount: 6,
        length: 4,
        oneOf: ["h", "l", "c", "s"],
        ignore: ["v", "z", "j", "t"],
        repeat: 2,
      });
      const result = getRandomWords(config.amount, config);
    });
    test("ignore - 2", () => {
      const config = getWordsFilterConfig({
        amount: 8,
        length: 4,
        oneOf: ["l"],
        ignore: ["v", "z", "j", "h"],
        repeat: 1,
      });
      const result = getRandomWords(config.amount, config);
      console.log("[random-data.test.ts,43] result: ", result);
      // console.log("[random-data.test.ts,43] result: ", result);
    });
    test("ignore - 3", () => {
      const config = getWordsFilterConfig({
        amount: 2,
        sequence: ["uckled"],
        oneOf: ["l"],
        anyOrder: [],
        // ignore: ['h' , 'o'],
        ignore: ["v", " z", " j", " h"],
        repeat: 1,
        length: undefined,
      });
      const result = getRandomWords(config.amount, config);
      console.log("[random-data.test.ts,43] result: ", result);
      // console.log("[random-data.test.ts,43] result: ", result);
    });
  });
});

describe("Order of config fields", () => {
  test("1", () => {
    const config = getWordsFilterConfig({
      amount: 4,
      length: 4,
      oneOf: ["r"],
      sequence: ["sh"],
    });
    const { words } = getRandomWords(config.amount, config);
    const [first, ...rest] = words;
    expect(first).toBe("rush");
    const okay = rest.every((word) => word.includes("r"));
    expect(okay).toBe(true);
  });
  test("2", () => {
    const config = getWordsFilterConfig({
      amount: 4,
      length: 4,
      sequence: ["sh"],
      oneOf: ["r"],
    });
    const { words } = getRandomWords(config.amount, config);
    const [first, ...rest] = words;
    expect(first).toBe("rush");
    const okay = rest.every((word) => word.includes("sh"));
    expect(okay).toBe(true);
  });
  test("3", () => {
    const config = getWordsFilterConfig({
      amount: 4,
      length: 5,
      sequence: ["sh"],
      oneOf: ["r"],
    });
    const { words } = getRandomWords(config.amount, config);
    const [first, ...rest] = words;
    expect(first).toHaveLength(5);
    const okay = rest.every((word) => word.includes("sh"));
    expect(okay).toBe(true);
  });
});

describe("Lessons", () => {
  describe("alphabet", () => {
    test("alphabet - 1", () => {
      const config = getWordsFilterConfig({
        lesson: "alphabet",
      });
      const { words } = getRandomWords(config.amount, config);
      expect(words[0]).toBe(ALPHABET);
    });
  });

  describe("alphabet-chunks", () => {
    test("alphabet - 1", () => {
      const config = getWordsFilterConfig({
        lesson: "alphabet-chunks",
        amount: 4,
        length: 4,
      });
      const { words } = getRandomWords(config.amount, config);
      expect(words.length).toBe(4);
      const lengthOkay = words.every((word) => word.length === 4);
      const chunkOkay = words.every((word) => ALPHABET.includes(word));
      expect(lengthOkay).toBe(true);
      expect(chunkOkay).toBe(true);
    });
  });
});
