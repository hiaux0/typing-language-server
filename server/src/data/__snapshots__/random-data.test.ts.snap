// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Normal getRandomWords 1`] = `
[
  "rush",
  "hire",
  "fair",
  "pray",
]
`;

exports[`Order of config fields 1 1`] = `
[
  "rush",
  "rule",
  "roof",
  "sour",
]
`;
